File: backend/models/listing_task.py
"""
Listing Task Pydantic models for property-specific tasks.

Following Context7 best practices for Pydantic v2 with Field validation.
"""

from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field
from enum import Enum


class TaskCategory(str, Enum):
    """Task categories."""
    ADMIN = "ADMIN"
    MARKETING = "MARKETING"
    PHOTO = "PHOTO"
    STAGING = "STAGING"
    INSPECTION = "INSPECTION"
    OTHER = "OTHER"


class TaskStatus(str, Enum):
    """Task status values."""
    OPEN = "OPEN"
    CLAIMED = "CLAIMED"
    IN_PROGRESS = "IN_PROGRESS"
    DONE = "DONE"
    FAILED = "FAILED"
    CANCELLED = "CANCELLED"


class VisibilityGroup(str, Enum):
    """Task visibility groups."""
    BOTH = "BOTH"
    AGENT = "AGENT"
    MARKETING = "MARKETING"


class ListingTaskBase(BaseModel):
    """Base listing task model with common fields."""
    listing_id: str = Field(..., description="Listing ID this task is for")
    realtor_id: Optional[str] = Field(None, description="Realtor for this listing")
    name: str = Field(..., min_length=1, max_length=500, description="Task name")
    description: Optional[str] = Field(None, max_length=5000, description="Task description")
    task_category: TaskCategory = Field(..., description="Task category")
    status: TaskStatus = Field(default=TaskStatus.OPEN, description="Task status")
    priority: int = Field(default=0, ge=0, le=10, description="Priority (0-10)")
    visibility_group: VisibilityGroup = Field(default=VisibilityGroup.BOTH, description="Who can see this task")
    assigned_staff_id: Optional[str] = Field(None, description="Staff member assigned to this task")
    due_date: Optional[datetime] = Field(None, description="Task due date")
    inputs: dict = Field(default_factory=dict, description="Task input parameters")
    outputs: dict = Field(default_factory=dict, description="Task output results")


class ListingTaskCreate(ListingTaskBase):
    """Model for creating a new listing task."""
    pass


class ListingTaskUpdate(BaseModel):
    """Model for updating listing task (all fields optional)."""
    name: Optional[str] = Field(None, min_length=1, max_length=500)
    description: Optional[str] = Field(None, max_length=5000)
    task_category: Optional[TaskCategory] = None
    status: Optional[TaskStatus] = None
    priority: Optional[int] = Field(None, ge=0, le=10)
    visibility_group: Optional[VisibilityGroup] = None
    assigned_staff_id: Optional[str] = None
    due_date: Optional[datetime] = None
    inputs: Optional[dict] = None
    outputs: Optional[dict] = None


class ListingTask(ListingTaskBase):
    """Complete listing task model (database record)."""
    task_id: str = Field(..., description="Unique identifier (ULID)")
    claimed_at: Optional[datetime] = Field(None, description="When task was claimed")
    completed_at: Optional[datetime] = Field(None, description="When task was completed")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")
    deleted_at: Optional[datetime] = Field(None, description="Soft delete timestamp")
    deleted_by: Optional[str] = Field(None, description="User who deleted task")

    model_config = {
        "from_attributes": True,
        "json_schema_extra": {
            "example": {
                "task_id": "01HWQK7A1B2CDEF4H5JKLMNOP",
                "listing_id": "01HWQK6Z9X8ABCD3F4G5HVWXY",
                "realtor_id": "01HWQK5Z8Y9ABCD3F4G5HVWXYZ",
                "name": "Schedule professional photography",
                "description": "Coordinate with photographer for property photos",
                "task_category": "PHOTO",
                "status": "OPEN",
                "priority": 5,
                "visibility_group": "BOTH",
                "assigned_staff_id": None,
                "due_date": "2024-02-01T14:00:00Z",
                "inputs": {},
                "outputs": {},
                "claimed_at": None,
                "completed_at": None,
                "created_at": "2024-01-15T10:30:00Z",
                "updated_at": "2024-01-15T10:30:00Z",
                "deleted_at": None,
                "deleted_by": None
            }
        }
    }


class ListingTaskListResponse(BaseModel):
    """Paginated list of listing tasks."""
    data: list[ListingTask] = Field(..., description="List of listing tasks")
    total: int = Field(..., ge=0, description="Total count")
    page: int = Field(..., ge=1, description="Current page number")
    page_size: int = Field(..., ge=1, le=100, description="Items per page")

    model_config = {
        "json_schema_extra": {
            "example": {
                "data": [],
                "total": 0,
                "page": 1,
                "page_size": 50
            }
        }
    }

File: apps/Operations Center/Packages/OperationsCenterKit/Sources/OperationsCenterKit/Models/ListingTask.swift
//
//  ListingTask.swift
//  Operations Center
//
//  Data model for listing_tasks table
//

import Foundation

public struct ListingTask: Identifiable, Codable, Sendable {
    public let id: String
    public let listingId: String
    public let realtorId: String?
    public let name: String
    public let description: String?
    public let taskCategory: TaskCategory
    public let status: TaskStatus
    public let priority: Int
    public let visibilityGroup: VisibilityGroup
    public let assignedStaffId: String?
    public let dueDate: Date?
    public let claimedAt: Date?
    public let completedAt: Date?
    public let createdAt: Date
    public let updatedAt: Date
    public let deletedAt: Date?
    public let deletedBy: String?
    public let inputs: [String: AnyCodable]?
    public let outputs: [String: AnyCodable]?

    // MARK: - Nested Types

    public enum TaskCategory: String, Codable, Sendable {
        case admin = "ADMIN"
        case marketing = "MARKETING"
        case photo = "PHOTO"
        case staging = "STAGING"
        case inspection = "INSPECTION"
        case other = "OTHER"
    }

    public enum TaskStatus: String, Codable, Sendable {
        case open = "OPEN"
        case claimed = "CLAIMED"
        case inProgress = "IN_PROGRESS"
        case done = "DONE"
        case failed = "FAILED"
        case cancelled = "CANCELLED"
    }

    public enum VisibilityGroup: String, Codable, Sendable {
        case both = "BOTH"
        case agent = "AGENT"
        case marketing = "MARKETING"
    }

    // MARK: - CodingKeys

    enum CodingKeys: String, CodingKey {
        case id = "task_id"
        case listingId = "listing_id"
        case realtorId = "realtor_id"
        case name
        case description
        case taskCategory = "task_category"
        case status
        case priority
        case visibilityGroup = "visibility_group"
        case assignedStaffId = "assigned_staff_id"
        case dueDate = "due_date"
        case claimedAt = "claimed_at"
        case completedAt = "completed_at"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case deletedAt = "deleted_at"
        case deletedBy = "deleted_by"
        case inputs
        case outputs
    }

    // MARK: - Computed Properties

    public var isCompleted: Bool {
        status == .done
    }

    public var isOverdue: Bool {
        guard let dueDate, status != .done else { return false }
        return dueDate < Date()
    }
}

// MARK: - AnyCodable Helper

/// Helper for encoding/decoding dynamic JSON
public struct AnyCodable: Codable, Sendable {
    public let value: Any

    public init(_ value: Any) {
        self.value = value
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()

        if let int = try? container.decode(Int.self) {
            value = int
        } else if let double = try? container.decode(Double.self) {
            value = double
        } else if let string = try? container.decode(String.self) {
            value = string
        } else if let bool = try? container.decode(Bool.self) {
            value = bool
        } else if let dict = try? container.decode([String: AnyCodable].self) {
            value = dict.mapValues { $0.value }
        } else if let array = try? container.decode([AnyCodable].self) {
            value = array.map { $0.value }
        } else if container.decodeNil() {
            value = NSNull()
        } else {
            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "Unable to decode value"
            )
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()

        switch value {
        case let int as Int:
            try container.encode(int)
        case let double as Double:
            try container.encode(double)
        case let string as String:
            try container.encode(string)
        case let bool as Bool:
            try container.encode(bool)
        case let dict as [String: Any]:
            let codableDict = dict.mapValues { AnyCodable($0) }
            try container.encode(codableDict)
        case let array as [Any]:
            let codableArray = array.map { AnyCodable($0) }
            try container.encode(codableArray)
        case is NSNull:
            try container.encodeNil()
        default:
            throw EncodingError.invalidValue(
                value,
                EncodingError.Context(
                    codingPath: container.codingPath,
                    debugDescription: "Unable to encode value"
                )
            )
        }
    }
}
